### PR TITLE
zsh completion

### PR DESCRIPTION
**Pull request by pseyfert**
proof of concept for zsh tab completion, a few todos, a few hacks. for discussion and stuff

### CODE REVIEW

**Comment by pseyfert** [comment]
`blinky -S <pkg1> <pkg2> --verb⇥` results in
```
pseyfert@robusta:~ > blinky -S <pkg1> <pkg2> --verbTraceback (most recent call last):
  File "/home/pseyfert/.local/bin/blinky", line 246, in <module>
    exp = re.compile(args.pkg_candidates[0] + '.*')
IndexError: list index out of range
```
i.e. I'm passing insane arguments to blinky

**Comment by pseyfert** [comment]
fixed. a bit surprised how that code path is invisible to `_message` debugging

File: completion/_blinky [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+  typeset -U packages
+  _wanted packages expl "installed non-repository packages" compadd "${(@)packages}"
+}
+
+_blinky_completions_all_packages() {
+  local -i minimum_package_name_length
+  # steer minimum package name length for launching package search
+  # set with:
+  # zstyle :completion:expand-word:complete:blinky:pkgcomp: numbers 4
```
Comment (pseyfert) [inline completion/_blinky:46]: @cherti how do you like this. By default `blinky -S abc⇥ ` will only complete if the current package name has at least four characters (i.e. not in this example). This is threshold by the user with
```zstyle :completion:expand-word:complete:blinky:pkgcomp: numbers 4```

**Comment by pseyfert** [comment]
ARGH, we need to review the combinations of options with commands. e.g. atm there is no completion for `-S --force-review`, although i see it's actually useful.

File: completion/_blinky (hunk 1 above)
Comment (pseyfert) [reply completion/_blinky:46]: ping @cherti I don't recall if we discussed this. any executive decision from your side?

The issue for me is that completion for `blinky -S <tab>` will always be on the slow side (sends a query to aur and parses the response, and even some overhead for launching a python process). otoh, not completing package names isn't great either as I would like to provide users with the functionality to just get long package names typo free by hammering tab. A threshold on a few letters avoids immediately freezing the shell by hammering tab before starting a package name (the list of all package names doesn't help a lot, one needs to type a few letters anyway. later, when there are only O(10) options, one can hammer tab or use arrows to get the right one, and can accept a second of delay easier as there's only "pick the right and hammer return" left to do, it interrupts the thought process by the user less, in my experience.)

Comment (cherti) [reply completion/_blinky:46]: I do think that this looks like a viable way forward. keeping a local list of packages (and therefore having the user to actually update them) doesn't sound too appealing to me, so I think this is a very good compromise, especially as it is configurable.

Comment (pseyfert) [reply completion/_blinky:46]: okay. suggestion where to document the configuration?

**Comment by pseyfert** [comment]
I rebased (and also rewrote the completion function, cleaning it up such that it is hopefully a lot easier to read, and also adding a few flags that were missing).

What I noticed is (and this is probably another point for executive diecision):

At the moment I try to be smart and don't offer double dash flags that won't have any effect (`-Ss` doesn't use `--difftool`). That is done by checking the `-S*` flag on the current command lne and filling the list of `--`options correspondingly. So far so good. If there is no `-S` on the current command line, then basically no `--` is suggested. One could flip the logic and suggest all `--` flags unless it is clear from the command line that they won't be needed. (effectively one would need to manipulate the 'else' branch in the case statement). It may be inconvenient that

```
blinky --diff<TAB>
```
doesn't complete. (NB: manipulation in the middle of the line `blinky --diff<TAB> -Sr` **does** work, i.e. the order of flags is not important for what I describe, only the presence).

**Comment by cherti** [comment]
> One could flip the logic and suggest all `--` flags unless it is clear from the command line that they won't be needed.

I think this is the more intuitive approach as the users would typically, when used to blinky, only use/tabcomplete flags that will work with the command they will finally use.

**Comment by pseyfert** [comment]
I just realised we didn't post the outcome of our irc chat here.
There are two situations relevant for my last comment:

- `-S<something>` is present on the command line → Only relevant `--<somethingelse>` flags get suggested. I.e. `blinky -S --diff<TAB>` completes and `blinky -Sr --diff` doesn't complete.

- `-S<something>` is not present on the command line → We have two options:

1. One could suggest all flags (i.e. flags that are valid for at least one command). I'd call this "suggest all flags that aren't ruled out already". In short `blinky --diff<TAB>` completes, `blinky --verb<TAB>` completes.
2. One suggests only those flags that are valid for all commands (in practice `--verbose`).  I'd call this "suggest only flags that will definitively make sense". In short `blinky --diff<TAB>` won't complete, `blinky --verb<TAB>` completes.

I have the completion on my system in production since a while (in setting 2) and I am confident I would prefer 1. (Yes, the grass is always greener on the other side, but also the total number of flags isn't that high, so spamming the terminal with irrelevant flags won't actually spam).

**Comment by cherti** [comment]
I also think 1 is clearly preferrable, as we do not have that many flags that this would be inconvenient.

**Comment by pseyfert** [comment]
Okay, done so in 45eb2c8a5f50e6e794b06408435ddae9162d56d4

**Comment by pseyfert** [comment]
I think we are almost good here, *except* we're about to invalidate the arguments in #12 . So I would say we update the completion once #12 is merged.

File: scripts/blinky [hunk 2]
```diff
@@ -12,6 +12,7 @@ primary.add_argument("-S", action='store_true', default=False, dest='install', h
 primary.add_argument("-Sr", action='store_true', default=False, dest='rebuildpkg', help="Reinstall package(s) from AUR (including rebuild)")
 primary.add_argument("-Srr", action='store_true', default=False, dest='rebuildtree', help="Reinstall package(s) and dependencies from AUR (including rebuild)")
 primary.add_argument("-Ss", action='store_true', default=False, dest='search', help="Search for package(s) in AUR")
+primary.add_argument("-complete", action='store_true', default=False, dest='complete', help="completion helper")
```
Comment (cherti) [inline scripts/blinky]: we should set help to `argparse.SUPPRESS` to make the completion-flag vanish from `--help`.

https://docs.python.org/3/library/argparse.html#help

Comment (pseyfert) [reply scripts/blinky]: TIL!

File: completion/_blinky [hunk 3]
```diff
@@ -0,0 +1,146 @@
+#compdef blinky
+# Copyright (C) 2018 Paul Seyfert <pseyfert.mathphys@gmail.com>
+#
+# This software is distributed under the terms of the GNU General Public
+# Licence version 3 (GPL Version 3), copied verbatim in the file "LICENSE".
+
+# heavily inspired by the _pacman completion
+
+local -a _blinky_opts_all _blinky_opts_install _blinky_opts_build
+
```
Comment (pseyfert) [inline completion/_blinky:10]: no visible: I anticipate the dropping of --local-path

File: completion/_blinky (hunk 1 above)
Comment (cherti) [reply completion/_blinky:46]: For now the Readme, or alternatively a file `docs/completion.md`, we do not have manpages yet, this is something we probably should do at some point before 1.0.

Comment (pseyfert) [reply completion/_blinky:46]: added to README

**Comment by pseyfert** [comment]
rebased. so assuming with the activity in the last days on the repository nothing broke (famous last words), this PR should be ready to merge. I'll put it in production on my side and will post when I see a showstopper.

File: completion/_blinky (hunk 3 above)
Comment (pseyfert) [reply completion/_blinky:10]: this was done in 7b8d8e92cfd8ea58a44599470f5f19b85ec98206 so no longer *anticipating*

**Comment by pseyfert** [comment]
ftr. didn't encounter issues on my side

**Comment by cherti** [comment]
So you consider this ready for merge, @pseyfert ?

**Comment by pseyfert** [comment]
yes

**Comment by cherti** [comment]
alright, then let's merge it! can you quickly reopen against branch `main`? shouldn't change anything functionally, we just had a branch renaming here, so just reopen the same PR against `main` and we can merge this. :)

**Comment by pseyfert** [comment]
target branch updated

**Pull request merged**